### DISCUSSION TITLE
How is an user's locale being stored?

### DISCUSSION THREAD

**Discussion by RaenonX** [amannn/next-intl#429] [Q&A] [answered]
Precisely, how is the return of `useLocale()` from `'next-intl'` being determined?

The reason I asked this is because - when I was developing my app, and using `useRouter()` from `next-intl/client` to push the locale, it works like a charm. However, when I put it to production and run my app using `next start` with `pm2` 8 instances instead, the locale switching failed.

Say I have pages `/a` and `/b` with locale `zh` and `en`. If I am on locale `en` at `/a`, and switch to locale `zh` on `/a` using `router.push()`, the locale of `/a` changes. However, if there's a hyperlink using `Link` from `next/link` which navigates to `/b`, if I click it, `/b` usually shows the page in `en` (which is the original locale) instead of `zh`.

I might be missing something here, so I am currentlyt using discussion instead of issues.

If the locale is something stored server-side and instance-dependent, does that mean I need to enable something to use `NEXT_LOCALE` instead?

Thanks!

**Comment by amannn** [comment, marked as answer]
The server side is stateless and changing the locale means that a cookie is set in the browser. Therefore any request from a browser that has received this cookie will use the latest locale.

Your situation sounds similar to https://next-intl-example-next-13.vercel.app/, which also has two routes. Does the behavior of the example look correct to you? It's hosted on Vercel, which is completely stateless.

**Reply by RaenonX** [reply]
This is what I want, but the issue is that I got similar language switching effect using `next dev`, but not `next start` under `pm2`.

Say I have page `/a` and `/b` with locale `zh` and `kr`. If I start from `/a` in `zh` and switch to `/a` in `kr`, then go to `/b`, either `/b` shows up in `zh` or the navigation to `/b` crashes (if I refresh, `/b` shows up in `zh`).

I can see `NEXT_LOCALE` being changed according to the locale change in `/a`, but once I navigate to `/b`, it reverts back.

https://github.com/RaenonX-PokemonSleep/pokemon-sleep-ui/blob/5ad0a82c8053f1f26766e957506bf247bb264efe/src/ui/base/navbar/languageSwitch/client.tsx#L17

This is the implementation of my current language switch, and this the link below is the current problematic production implementation.

https://pks.raenonx.cc/en/cooking

**Reply by amannn** [reply]
You might want to have a look at how the navigation is implemented in next-intl-example-next-13 (using `next-intl/link`), that might help!

**Reply by RaenonX** [reply]
I think `next-intl/link` might actually be the caveat here, let me try it. Thanks!

Also, is it better if I replace all `import Link from 'next/link'` usages to `import Link from 'next-intl/link'`?

**Reply by amannn** [reply]
`next-intl/link` automatically handles the locale, so we recommend that, yes!

**Reply by RaenonX** [reply]
The line below is `href="/"`, but the demo website https://next-intl-example-next-13.vercel.app/ doesn't change the current pathname, did I miss something here? My observation is that `href` needs to be the pathname from `usePathname()` of `next-intl` to do the same page transition.

https://github.com/amannn/next-intl-example-rsc/blob/8a7151ce9741e1b34d651ed5c22b823d93d809c8/src/components/LocaleSwitcher.tsx#L9

**Reply by amannn** [reply]
Here's the source code of https://next-intl-example-next-13.vercel.app/ : https://github.com/amannn/next-intl/tree/main/examples/example-next-13

**Reply by RaenonX** [reply]
Seems like it worked and I was looking at the wrong repo, thanks!
